File: Python version/platforms/desktop/desktop_service.py
from base.service import Service
from abc import abstractmethod
from base.option_descriptor import OptionDescriptor
from base.input_program import InputProgram
from base.output import Output
import subprocess
import time
from threading import Thread


class DesktopService(Service):
    """Specialization for a desktop platform."""

    def __init__(self, exe_path):
        self._exe_path = exe_path  # Stores solver's executable path
        # Stores option string in order to enable solver to read from standard input
        self._load_from_stdin_option = None

    def get_exe_path(self):
        """Return execution path of DesktopService."""
        return self._exe_path

    @abstractmethod
    def _get_output(self, output, error):
        pass

    def set_exe_path(self, exe_path):
        """Set _exe_path to a new path.

        The parameter exe_path is a string representing the path for the
        new solver.
        """
        self._exe_path = exe_path

    def start_async(self, callback, programs, options):
        """Start a new process for the _exe_path and starts solving
        asyncronously."""
        class MyThread(Thread):
            def __init__(self, start_sync):
                Thread.__init__(self)
                self.start_sync = start_sync

            def run(self):
                callback.callback(self.start_sync(programs, options))

        th = MyThread(self.start_sync)
        th.start()

    def start_sync(self, programs, options):
        """Start a new process for the _exe_path and starts solving
        syncronously."""
        option = ""
        for o in options:
            if o is not None:
                option += o.get_options()
                option += o.get_separator()
            else:
                print("Warning : wrong " +
                      str(OptionDescriptor().__class__.__name__))
                
        files_paths = list()
        final_program = ""
        for p in programs:
            if p is not None:
                final_program += p.get_programs()
                for path in p.get_files_paths():
                    if len(path) != 0:
                        files_paths.append(path)
            else:
                print("Warning : wrong " +
                      str(InputProgram().__class__.__name__))

        if self._exe_path is None:
            return Output("", "Error: executable not found")

        exep = str(self._exe_path)
        opt = str(option)

        lis = list()
        lis.append(exep)
        if opt != "":
            lis.append(opt)
        for path in files_paths:
            lis.append(path)
        if self._load_from_stdin_option != "" and final_program != "":
            lis.append(self._load_from_stdin_option)

        print(exep + " ", end='')
        if opt != "":
            print(opt + " ", end='')
        for path in files_paths:
            print(path + " ", end='')
        if final_program != "":
            print(self._load_from_stdin_option)
        else:
            print()
            
        start = int(time.time() * 1e+9)

        proc = subprocess.Popen(
            lis,
            universal_newlines=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.PIPE,
        )

        output, error = proc.communicate(final_program)

        end = int(time.time() * 1e+9)

        print("Total time : " + str(end - start))
        print("")

        return self._get_output(output, error)

File: Java version/app/src/main/java/it/unical/mat/embasp/platforms/desktop/DesktopService.java
package it.unical.mat.embasp.platforms.desktop;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.List;

import it.unical.mat.embasp.base.Callback;
import it.unical.mat.embasp.base.InputProgram;
import it.unical.mat.embasp.base.OptionDescriptor;
import it.unical.mat.embasp.base.Output;
import it.unical.mat.embasp.base.Service;

/**
 * is a specialization for a Desktop platform
 *
 * @see Service
 */

public abstract class DesktopService implements Service {
	/** Stores solver's executable path */
	protected String exe_path;
	protected String load_from_STDIN_option;

	public DesktopService(final String exe_path) {
		this.exe_path = exe_path;
	}

	public String getExePath() {
		return exe_path;
	}

	abstract protected Output getOutput(String output, String error);

	/**
	 * set {@link #exe_path} to a new path*
	 *
	 * @param exe_path
	 *            a string representing the path for the new solver
	 */
	public void setExePath(final String exe_path) {
		this.exe_path = exe_path;
	}

	/**
	 * Start a new process for the {@link #exe_path} and starts solving
	 *
	 * @see it.unical.mat.embasp.base.Service#startAsync(Callback, List, List)
	 */
	@Override
	public void startAsync(final Callback callback, final List<InputProgram> programs, final List<OptionDescriptor> options) {

		new Thread() {
			@Override
			public void run() {
				callback.callback(startSync(programs, options));
			}
		}.start();

	}

	/**
	 * Start a new process for the {@link #exe_path} and starts solving
	 *
	 * @see it.unical.mat.embasp.base.Service#startSync(List, List)
	 */
	@Override
	public Output startSync(final List<InputProgram> programs, final List<OptionDescriptor> options) {

		String option = new String();
		for (final OptionDescriptor o : options)
			if (o != null) {
				option += o.getOptions();
				option += o.getSeparator();
			} else
				System.err.println("Warning : wrong " + OptionDescriptor.class.getName());

		String files_paths = new String();
		String final_program = new String();

		for (final InputProgram p : programs)
			if (p != null) {
				final_program += p.getPrograms();
				for(final String program_file: p.getFilesPaths()){
					File f = new File(program_file);
					if(f.exists() && !f.isDirectory()) { 
						files_paths += program_file;
						files_paths += " ";
					}
					else
						System.err.println("Warning : the file " + f.getAbsolutePath() + " does not exists.");
				}
			} else
				System.err.println("Warning : wrong " + InputProgram.class.getName());

		final StringBuffer solverOutput = new StringBuffer();
		final StringBuffer solverError = new StringBuffer();

		try {

			final long startTime = System.nanoTime();
			
			File tmpFile = null;

			final StringBuffer stringBuffer = new StringBuffer();
			if (exe_path == null)
				return new Output("", "Error: executable not found");
			stringBuffer.append(exe_path).append(" ").append(option).append(" ").append(files_paths);
			if (!final_program.isEmpty()){
				tmpFile=writeToFile("tmp", final_program);
				stringBuffer.append(" ").append(tmpFile.getAbsolutePath());
			}

			System.err.println(stringBuffer.toString());
			final Process solver_process = Runtime.getRuntime().exec(stringBuffer.toString());

			Thread threadOutput=new Thread() {
				@Override
				public void run() {
					try {

						final BufferedReader bufferedReaderOutput = new BufferedReader(new InputStreamReader(solver_process.getInputStream()));

						// Read output of the solver and store in solverOutput
						String currentLine;
						while ((currentLine = bufferedReaderOutput.readLine()) != null)
							solverOutput.append(currentLine + "\n");
					} catch (final IOException e) {
						e.printStackTrace();
					}

				}
			};
			threadOutput.start();
			threadOutput.join();
			
			Thread threadError = new Thread() {
				@Override
				public void run() {
					try {

						final BufferedReader bufferedReaderError = new BufferedReader(new InputStreamReader(solver_process.getErrorStream()));

						String currentErrLine;
						while ((currentErrLine = bufferedReaderError.readLine()) != null)
							solverError.append(currentErrLine + "\n");

					} catch (final IOException e) {
						e.printStackTrace();
					}

				}
			};
			threadError.start();
			threadError.join();
			
			final PrintWriter writer = new PrintWriter(solver_process.getOutputStream());
			writer.println(final_program);
			if (writer != null)
				writer.close();

			solver_process.waitFor();

			final long stopTime = System.nanoTime();
			System.err.println("Total time : " + (stopTime - startTime));
			
			if(tmpFile!=null) tmpFile.delete();
			
			return getOutput(solverOutput.toString(), solverError.toString());

		} catch (final IOException e2) {
			e2.printStackTrace();
		} catch (final InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

		return getOutput("", "");

	}
	
	protected File writeToFile(String pFilename, String sb) throws IOException {
	    File tempDir = new File(System.getProperty("java.io.tmpdir"));
	    File tempFile = File.createTempFile(pFilename, ".tmp", tempDir);
	    FileWriter fileWriter = new FileWriter(tempFile, true);
	    BufferedWriter bw = new BufferedWriter(fileWriter);
	    bw.write(sb);
	    bw.close();
	    return tempFile;
	}

}
